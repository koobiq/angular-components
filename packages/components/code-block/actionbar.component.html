<div class="kbq-code-block__fade-gradient"></div>

<button
    *ngIf="multiLine"
    kbq-button
    [color]="'contrast'"
    [kbqStyle]="'transparent'"
    [kbqTooltip]="softWrap ? config.softWrapOffTooltip : config.softWrapOnTooltip"
    (click)="toggleSoftWrap.emit()"
>
    <i
        *ngIf="!softWrap"
        kbq-icon="mc-word-wrap_16"
    ></i>
    <i
        *ngIf="softWrap"
        kbq-icon="mc-word-wrap-slash_16"
    ></i>
</button>

<button
    *ngIf="canLoad"
    kbq-button
    [color]="'contrast'"
    [kbqStyle]="'transparent'"
    [kbqTooltip]="config.downloadTooltip"
    (click)="downloadCode.emit()"
>
    <i kbq-icon="mc-download_16"></i>
</button>

<button
    #copyTooltip="kbqTooltip"
    kbq-button
    [color]="'contrast'"
    [kbqStyle]="'transparent'"
    [kbqTooltip]="copyTooltipText"
    (click)="onCopy()"
>
    <i kbq-icon="mc-copy-o_16"></i>
</button>

<button
    *ngIf="codeFiles[selectedTabIndex].link"
    kbq-button
    [color]="'contrast'"
    [kbqStyle]="'transparent'"
    [kbqTooltip]="config.openExternalSystemTooltip"
    (click)="openExternalSystem.emit()"
>
    <i kbq-icon="mc-external-link_16"></i>
</button>
