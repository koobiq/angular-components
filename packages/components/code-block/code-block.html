@let file = files[activeFileIndex];

<div class="kbq-code-block__header">
    @if (!hideTabs) {
        <div class="kbq-code-block__header-tabs">
            <div kbqTabNavBar transparent [tabNavPanel]="tabPanel">
                @for (file of files; track file; let index = $index) {
                    <div
                        kbqTabLink
                        [active]="activeFileIndex === index"
                        [disabled]="files.length === 1"
                        (click)="onSelectedTabChange(index)"
                    >
                        {{ file.filename || fallbackFileName }}
                    </div>
                }
            </div>
        </div>
    }

    <div class="kbq-code-block__actionbar">
        @if (hideTabs) {
            <div class="kbq-code-block__actionbar__fade-gradient"></div>
        }

        <div class="kbq-code-block__actionbar__button-stack">
            @if (canToggleSoftWrap) {
                <button
                    kbq-button
                    class="kbq-code-block__actionbar__soft-wrap-button"
                    [color]="componentColor.Contrast"
                    [kbqStyle]="buttonStyle.Transparent"
                    [kbqTooltip]="
                        softWrap ? localeConfiguration.softWrapOffTooltip : localeConfiguration.softWrapOnTooltip
                    "
                    (click)="toggleSoftWrap()"
                >
                    @if (softWrap) {
                        <i kbq-icon="kbq-wrap-text-slash_16"></i>
                    } @else {
                        <i kbq-icon="kbq-wrap-text_16"></i>
                    }
                </button>
            }

            @if (canDownload) {
                <button
                    kbq-button
                    class="kbq-code-block__actionbar__download-button"
                    [color]="componentColor.Contrast"
                    [kbqStyle]="buttonStyle.Transparent"
                    [kbqTooltip]="localeConfiguration.downloadTooltip"
                    (click)="downloadCode()"
                >
                    <i kbq-icon="kbq-arrow-down-to-line_16"></i>
                </button>
            }

            @if (canCopy) {
                <div #copyButtonTooltip="kbqTooltip" [kbqTooltip]="localeConfiguration.copyTooltip">
                    <button
                        kbq-button
                        class="kbq-code-block__actionbar__copy-button"
                        [color]="componentColor.Contrast"
                        [kbqStyle]="buttonStyle.Transparent"
                        (click)="copyCode()"
                    >
                        <i kbq-icon="kbq-file-multiple-o_16"></i>
                    </button>
                </div>
            }

            @if (file.link) {
                <button
                    kbq-button
                    class="kbq-code-block__actionbar__link-button"
                    [color]="componentColor.Contrast"
                    [kbqStyle]="buttonStyle.Transparent"
                    [kbqTooltip]="localeConfiguration.openExternalSystemTooltip"
                    (click)="openLink()"
                >
                    <i kbq-icon="kbq-arrow-up-right-from-square_16"></i>
                </button>
            }
        </div>
    </div>
</div>

<pre
    #tabPanel="kbqTabNavPanel"
    cdkMonitorElementFocus
    cdkScrollable
    kbqTabNavPanel
    class="kbq-code-block__content kbq-scrollbar"
    [style.max-height.px]="calculatedMaxHeight"
    [tabIndex]="codeContentTabIndex"
><code kbqCodeBlockHighlight class="kbq-code-block__code" [file]="file"></code></pre>

@if (maxHeight) {
    <div
        class="kbq-code-block__view-all"
        [class.kbq-code-block__view-all_collapsed]="!viewAll"
        [class.kbq-code-block__view-all_expanded]="viewAll"
    >
        <div class="kbq-code-block__view-all__wrapper">
            <button
                kbq-button
                class="kbq-code-block__view-all__button"
                [color]="componentColor.Theme"
                [kbqStyle]="buttonStyle.Transparent"
                (click)="toggleViewAll()"
                (keydown.enter)="onViewAllEnterKeydown()"
            >
                @if (viewAll) {
                    <i kbq-icon="kbq-chevron-up_16"></i>
                    {{ localeConfiguration.viewLessText }}
                } @else {
                    <i kbq-icon="kbq-chevron-down_16"></i>
                    {{ localeConfiguration.viewAllText }}
                }
            </button>
        </div>
    </div>
}
