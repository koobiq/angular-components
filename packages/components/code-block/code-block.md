Компонент показывает переформатированный текстовый контент и подсвечивает синтаксис.

<!-- example(code-block-common) -->
<!-- https://www.figma.com/file/Y5NWlgM9FDbdf7plybXvZK/Mosaic-Spec?node-id=2255%3A59388&t=8ZxsAMNJttUp2DMD-1 -->

### Описание работы и возможности

#### Нумерация строк

Пронумеровать строки полезно, чтобы дать ссылку на конкретное место в документе. Нумерация строк отключена по умолчанию.

<!-- example(code-block-line-numbers) -->

#### Изменение размера по содержимому

Блок может растягиваться и занимать доступную область экрана, если изучение кода — это одна из главных задач в интерфейсе. Когда место на экране ограничено, или код только дополняет основную информацию, проектировщик может использовать блок с фиксированным размером.

<!-- example(code-block-stretch) -->

Короткая строка кода без переноса:

<!-- example(code-block-single-line) -->

#### Свернуть или открыть полную версию

Для компактности бывает полезно показать часть документа с возможностью развернуть все. Блок увеличится по высоте и весь код будет виден — так удастся избавиться от двойной прокрутки. В свернутом виде нет прокрутки, недоступна функция смена режима переноса. Если сменить режим переноса строк, а потом свернуть-развернуть код, то настройка сохранится.

<!-- example(code-block-cut) -->

#### Мягкий перенос строк

По умолчанию строки не переносятся. Режим переноса в блоке можно настроить заранее, а также дать пользователю возможность управлять этим с помощью кнопки-переключателя в панели действий.

Иконка в кнопке означает будущее состояние после активации, иконка меняется после смены режима переноса.

<!-- example(code-block-line-wrap) -->

#### Показ нескольких документов

Несколько документов размещаются в блоке во вкладках. Заголовками вкладок могут имена документов, расширения файлов или другие подписи на усмотрение проектировщика.

<!-- example(code-block-tabs) -->

Блок заголовков вкладок отображается без тени по умолчанию. Если содержимое выходит за пределы области, то при скроллинге содержимого блока кода появляется тень под блоком заголовков вкладок.

<!-- example(code-block-tabs-with-overflow) -->

#### Заголовок блока

Если заголовок указан в конфигурации, то он будет показан над в шапке кодом, в той же панели располагаются и кнопки действий.

Заголовки надо обязательно указать в режиме просмотра нескольких документов, чтобы различать их. Заголовки будут показаны на вкладках.

<!-- example(code-block-title) -->

#### Дополнительные действия

Панель действий содержит кнопки:

-   Включить или выключить перенос строк
-   Скачать файл
-   Копировать текст документа в буфер обмена
-   Открыть во внешней системе. Например, редактировать код в Stackblitz. Ссылка открывается в новой вкладке.

Панель действий расположена в правом углу блока, видна по ховеру на блок или при фокусе на одну из кнопок, залипает при прокрутке. Конфигурация компонента определяет, какие действия будут доступны. Например, можно запретить пользователю менять режим переноса.

#### Сохранение документа

Пользователь может сохранить документ в файл. При настройке компонента можно указать специальное имя и расширение файла. Имя файла по умолчанию — «code». Расширение файла формируется по типу подсветки синтаксиса или .TXT, если не задано.

### Визуальное оформление

Предусмотрено два стиля оформление блока: с рамкой или с контрастной к фону страницы подложкой. Контрастный фон не дает блоку потеряться на экране с разномастным содержимым.

<!-- example(code-block-styling) -->

#### Без рамок на весь контейнер

Когда блок кода должен занять весь контейнер или экран, то лучше использовать оформление без рамки и фона. В примере открыта боковая панель, а в ней показан код:

<!-- example(code-block-noborder) -->
