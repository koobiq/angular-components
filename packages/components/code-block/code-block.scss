@use 'sass:meta';
@use 'sass:map';
@use '../core/styles/tokens';
@use '../core/styles/common/tokens' as *;


$tokens: meta.module-variables(tokens) !default;

$border-radius: kbq-css-variable(code-block-size-container-border-radius);
$border-width: kbq-css-variable(code-block-size-container-border-width);

/* stylelint-disable no-descending-specificity */
.kbq-code-block {
    display: block;
    position: relative;
    overflow: hidden;

    box-sizing: border-box;

    hyphens: none;
    -ms-hyphens: none;
    -moz-hyphens: none;
    -webkit-hyphens: none;

    border-radius: $border-radius;
    border-width: 0;

    .kbq-tab-header {
        // 4 buttons + margin
        $buttons-margin: calc(32 * 4) + 24px;

        padding:
            kbq-css-variable(code-block-size-header-padding-vertical)
            calc(var(
                --kbq-code-block-size-header-padding-right,
                map.get($tokens, code-block-size-header-padding-right)
            ) + $buttons-margin)
            kbq-css-variable(code-block-size-header-padding-vertical)
            kbq-css-variable(code-block-size-header-padding-left);

        border-width: $border-width $border-width 0 $border-width;
        border-style: solid;
        border-radius: $border-radius $border-radius 0 0;
    }

    &:has(.kbq-code-block__show-more) {
        .kbq-code-block__code {
            & > code {
                padding-bottom: kbq-css-variable(size-3xl) !important;
                overflow-x: scroll;
            }
        }
    }

    .kbq-code-block__code {
        > code {
            padding:
                kbq-css-variable(code-block-size-with-header-content-padding-top)
                kbq-css-variable(code-block-size-with-header-content-padding-horizontal)
                kbq-css-variable(code-block-size-with-header-content-padding-bottom)
                kbq-css-variable(code-block-size-with-header-content-padding-horizontal);

            border-radius: $border-radius;

            &:focus-visible {
                outline: none;
            }
        }

        .hljs-ln-numbers {
            padding-right: kbq-css-variable(code-block-size-with-header-content-content-gap-horizontal);
        }
    }

    &.kbq-code-block_hide-line-numbers .hljs-ln-line.hljs-ln-numbers {
        display: none;
    }

    &.kbq-code-block_single-file {
        .kbq-tab-label {
            &:before {
                content: none;
            }

            &:hover {
                background: transparent !important;
            }
        }
    }

    &:not(.kbq-code-block_no-header) {
        .kbq-tab-body__wrapper {
            border-top-width: 0;
        }
    }

    &.kbq-code-block_no-header {
        .kbq-tab-header {
            display: none;
        }

        .kbq-tab-body__wrapper {
            border-width: $border-width;
            border-radius: $border-radius;
        }

        .kbq-code-block__code {
            position: relative;

            & > code {
                padding:
                    kbq-css-variable(code-block-size-no-header-content-padding-vertical)
                    kbq-css-variable(code-block-size-no-header-content-padding-horizontal);
            }

            .hljs-ln-numbers {
                padding-right: kbq-css-variable(code-block-size-no-header-content-content-gap-horizontal);
            }
        }
    }

    &:not(:has(.kbq-code-block__show-more)) {
        .kbq-code-block__code {
            overflow-y: auto;
        }
    }

    pre {
        margin: 0;
    }

    .kbq-tab-body__wrapper {
        flex-direction: column;
        position: unset;

        border-width: $border-width;
        border-style: solid;
        border-radius: 0 0 $border-radius $border-radius;
    }

    .kbq-tab-body.kbq-tab-body__active {
        position: unset;
    }
}

.kbq-code-block__code {
    overflow-y: hidden;

    &.kbq-code-block__code_view-all {
        max-height: unset;
        overflow-y: auto;
    }

    &.kbq-code-block__code_soft-wrap {
        white-space: pre-wrap
    }

    .hljs-ln {
        border-collapse: unset;

        .hljs {
            padding: 0;
        }
    }

    .hljs-ln-numbers {
        vertical-align: baseline;
        text-align: right;
    }

    .hljs-ln-n:before {
        white-space: nowrap;
    }
}

.kbq-code-block__show-more {
    display: flex;
    justify-content: center;
    overflow: hidden;

    width: calc(100% - $border-width * 2);

    position: absolute;
    bottom: $border-width;
    border-radius: $border-radius;

    &.kbq-code-block__show-more_collapsed {
        padding-top: kbq-css-variable(code-block-size-collapse-collapsed-padding-top);

        padding-bottom: kbq-css-variable(code-block-size-collapse-collapsed-padding-bottom);
    }

    &.kbq-code-block__show-more_expanded {
        padding-top: kbq-css-variable(code-block-size-collapse-expanded-padding-top);

        margin-bottom: kbq-css-variable(code-block-size-collapse-expanded-padding-bottom);
    }

    .bg-wrapper {
        border-radius: kbq-css-variable(button-size-border-radius);
    }
}
/* stylelint-enable no-descending-specificity */
