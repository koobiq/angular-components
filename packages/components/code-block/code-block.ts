import { A11yModule, FocusMonitor } from '@angular/cdk/a11y';
import { Clipboard } from '@angular/cdk/clipboard';
import { Platform } from '@angular/cdk/platform';
import { CdkScrollable, CdkScrollableModule } from '@angular/cdk/scrolling';
import { DOCUMENT } from '@angular/common';
import {
    AfterViewInit,
    booleanAttribute,
    ChangeDetectionStrategy,
    ChangeDetectorRef,
    Component,
    DestroyRef,
    ElementRef,
    EventEmitter,
    inject,
    InjectionToken,
    Input,
    numberAttribute,
    Output,
    Provider,
    Renderer2,
    SecurityContext,
    ViewChild,
    ViewEncapsulation
} from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { DomSanitizer } from '@angular/platform-browser';
import { KbqButtonModule, KbqButtonStyles } from '@koobiq/components/button';
import {
    KBQ_LOCALE_SERVICE,
    KbqCodeBlockLocaleConfiguration,
    KbqComponentColors,
    ruRULocaleData
} from '@koobiq/components/core';
import { KbqIconModule } from '@koobiq/components/icon';
import { KbqTabsModule } from '@koobiq/components/tabs';
import { KbqToolTipModule, KbqTooltipTrigger } from '@koobiq/components/tooltip';
import { debounceTime, fromEvent, merge } from 'rxjs';
import { KbqCodeBlockHighlight } from './code-block-highlight';
import { KbqCodeBlockFile } from './types';

/** Localization configuration provider. */
export const KBQ_CODE_BLOCK_LOCALE_CONFIGURATION = new InjectionToken<KbqCodeBlockLocaleConfiguration>(
    'KBQ_CODE_BLOCK_LOCALE_CONFIGURATION',
    { factory: () => ruRULocaleData.codeBlock }
);

/** Utility provider for `KBQ_CODE_BLOCK_LOCALE_CONFIGURATION`. */
export const kbqCodeBlockLocaleConfigurationProvider = (configuration: KbqCodeBlockLocaleConfiguration): Provider => ({
    provide: KBQ_CODE_BLOCK_LOCALE_CONFIGURATION,
    useValue: configuration
});

/** Fallback file name for code block if file name is not specified. */
export const KBQ_CODE_BLOCK_FALLBACK_FILE_NAME = new InjectionToken<string>('KBQ_CODE_BLOCK_FALLBACK_FILE_NAME', {
    factory: () => 'code'
});

/** Utility provider for `KBQ_CODE_BLOCK_FALLBACK_FILE_NAME`. */
export const kbqCodeBlockFallbackFileNameProvider = (fileName: string): Provider => ({
    provide: KBQ_CODE_BLOCK_FALLBACK_FILE_NAME,
    useValue: fileName
});

/**
 * Component which highlights blocks of code.
 */
@Component({
    standalone: true,
    imports: [
        KbqTabsModule,
        KbqButtonModule,
        KbqCodeBlockHighlight,
        A11yModule,
        CdkScrollableModule,
        KbqToolTipModule,
        KbqIconModule
    ],
    selector: 'kbq-code-block',
    exportAs: 'kbqCodeBlock',
    templateUrl: './code-block.html',
    styleUrls: ['./code-block.scss', './code-block-tokens.scss'],
    host: {
        class: 'kbq-code-block',
        '[class.kbq-code-block_filled]': 'filled',
        '[class.kbq-code-block_outline]': '!filled',
        '[class.kbq-code-block_hide-line-numbers]': '!lineNumbers',
        '[class.kbq-code-block_hide-tabs]': 'hideTabs',
        '[class.kbq-code-block_no-border]': 'noBorder || filled',
        '[class.kbq-code-block_soft-wrap]': 'softWrap',
        '[class.kbq-code-block_view-all]': 'viewAll'
    },
    changeDetection: ChangeDetectionStrategy.OnPush,
    encapsulation: ViewEncapsulation.None
})
export class KbqCodeBlock implements AfterViewInit {
    @ViewChild('copyButtonTooltip') private readonly copyButtonTooltip?: KbqTooltipTrigger;
    @ViewChild(CdkScrollable) readonly scrollableCodeContent: CdkScrollable;

    /** Whether to display line numbers. */
    @Input({ transform: booleanAttribute }) lineNumbers = false;

    /** Whether the code block should be filled. */
    @Input({ transform: booleanAttribute }) filled: boolean = false;

    /** Added soft wrap toggle button.  */
    @Input({ transform: booleanAttribute }) canToggleSoftWrap: boolean = false;

    /** Whether sequences of whitespace should be preserved. */
    @Input({ transform: booleanAttribute }) softWrap: boolean = false;

    /** Event emitted when `softWrap` is changed. */
    @Output() readonly softWrapChange = new EventEmitter<boolean>();

    /**
     * Allows to view all the code, otherwise it will be hidden.
     * Works only with `maxHeight` property.
     */
    @Input({ transform: booleanAttribute }) viewAll: boolean = false;

    /** Event emitted when `viewAll` is changed. */
    @Output() readonly viewAllChange = new EventEmitter<boolean>();

    /**
     * Maximum height of the code block content, other parts will be hidden.
     * Can be toggled by `viewAll` property.
     */
    @Input({ transform: numberAttribute }) maxHeight: number;

    protected get calculatedMaxHeight(): number | null {
        return this.maxHeight > 0 && !this.viewAll ? this.maxHeight : null;
    }

    /** @deprecated Will be removed in next major release, use `canDownload` instead. */
    @Input({ transform: booleanAttribute })
    set canLoad(value: boolean) {
        this.canDownload = value;
    }

    /** Added download code button. */
    @Input({ transform: booleanAttribute }) canDownload: boolean = false;

    /** Added copy code button. */
    @Input({ transform: booleanAttribute }) canCopy: boolean = true;

    /** @deprecated Will be removed in next major release, use `files` instead. */
    @Input()
    set codeFiles(files: KbqCodeBlockFile[]) {
        this.files = files;
    }

    /**
     * @TODO Mark as `required`, after removing `codeFiles`
     *
     * Files to display.
     */
    @Input()
    set files(files: KbqCodeBlockFile[]) {
        this._files = files;

        if (this._files.length < this.activeFileIndex) {
            this.onSelectedTabChange(0);
        }

        if (this._files.length === 1 && !this._files[0].filename) {
            this.hideTabs = true;
        }
    }

    get files(): KbqCodeBlockFile[] {
        return this._files;
    }

    private _files: KbqCodeBlockFile[] = [];

    /** Defines which file (index) is active. */
    @Input({ transform: numberAttribute }) activeFileIndex = 0;

    @Output() readonly activeFileIndexChange = new EventEmitter<number>();

    /** Whether to hide border. */
    @Input({ transform: booleanAttribute }) noBorder: boolean = false;

    /**
     * Whether to hide header tabs.
     * Always `true` if there is only one file without filename.
     * Makes actionbar floating if tabs are hidden.
     */
    @Input({ transform: booleanAttribute })
    set hideTabs(value: boolean) {
        this._hideTabs = value;

        this.setupActionbarDisplay();
    }

    get hideTabs(): boolean {
        return this._hideTabs;
    }

    private _hideTabs: boolean = false;

    /** Component locale configuration. */
    protected get localeConfiguration(): KbqCodeBlockLocaleConfiguration {
        return this._localeConfiguration;
    }

    private _localeConfiguration: KbqCodeBlockLocaleConfiguration = inject(KBQ_CODE_BLOCK_LOCALE_CONFIGURATION);

    /** Code content tab index. */
    protected get codeContentTabIndex(): number {
        return this.canCodeContentBeFocused ? 0 : -1;
    }

    /**
     * Determines whether the code content can be focused.
     *
     * This checks if the scrollable code content element is present,
     * has a scroll, and the calculated maximum height is not set.
     */
    private get canCodeContentBeFocused(): boolean {
        const element = this.scrollableCodeContent?.getElementRef().nativeElement;
        return element && this.hasScroll(element) && !this.calculatedMaxHeight;
    }

    protected readonly componentColor = KbqComponentColors;
    protected readonly buttonStyle = KbqButtonStyles;

    private readonly elementRef = inject(ElementRef);
    private readonly changeDetectorRef = inject(ChangeDetectorRef);
    private readonly localeService = inject(KBQ_LOCALE_SERVICE, { optional: true });
    private readonly destroyRef = inject(DestroyRef);
    private readonly renderer = inject(Renderer2);
    private readonly platform = inject(Platform);
    private readonly focusMonitor = inject(FocusMonitor);
    private readonly clipboard = inject(Clipboard);
    private readonly domSanitizer = inject(DomSanitizer);
    private readonly document = inject<Document>(DOCUMENT);
    protected readonly fallbackFileName = inject(KBQ_CODE_BLOCK_FALLBACK_FILE_NAME);

    constructor() {
        this.localeService?.changes.pipe(takeUntilDestroyed()).subscribe(this.updateLocaleParams);
    }

    ngAfterViewInit(): void {
        this.handleScroll();
        this.handleHover();

        // Setup initial actionbar display state
        this.setupActionbarDisplay();

        this.copyButtonTooltip?.visibleChange.pipe(takeUntilDestroyed(this.destroyRef)).subscribe((isVisible) => {
            if (isVisible) {
                this.copyButtonTooltip!.content = this.localeConfiguration.copyTooltip;
            }
        });
    }

    /**
     * Toggles the `viewAll` property.
     *
     * When `viewAll` is set to `true`, the content of the code block will be
     * displayed in its entirety. When set to `false`, the content will be
     * limited to the `maxHeight` property. If the content exceeds the
     * `maxHeight`, it will be scrolled to the top and the `viewAll` button
     * will be displayed.
     */
    toggleViewAll(): void {
        this.viewAll = !this.viewAll;

        if (!this.viewAll) {
            this.scrollableCodeContent.scrollTo({ top: 0, behavior: 'instant' });
        }

        this.viewAllChange.emit(this.viewAll);
    }

    /**
     * Toggles `softWrap` property.
     *
     * When `softWrap` is set to `true`, the content of the code block will be
     * wrapped if it exceeds the height of the component. When set to `false`
     * the content will not be wrapped.
     */
    toggleSoftWrap(): void {
        this.softWrap = !this.softWrap;
        this.softWrapChange.emit(this.softWrap);
    }

    /**
     * Handles the change of the selected tab by updating the active file index
     * and scrolling to the top of the scrollable content.
     *
     * @param index - The index of the newly selected tab.
     */
    protected onSelectedTabChange(index: number): void {
        if (this.activeFileIndex !== index) {
            this.activeFileIndex = index;
            this.activeFileIndexChange.emit(this.activeFileIndex);
            this.scrollableCodeContent.scrollTo({ top: 0, behavior: 'instant' });
        }
    }

    /**
     * Handles the hover event on the code block element and updates the actionbar display accordingly.
     */
    private handleHover(): void {
        merge(
            fromEvent<MouseEvent>(this.elementRef.nativeElement, 'mouseenter'),
            fromEvent<MouseEvent>(this.elementRef.nativeElement, 'mouseleave')
        )
            .pipe(debounceTime(100), takeUntilDestroyed(this.destroyRef))
            .subscribe((event) => {
                this.setupActionbarDisplay(event?.type === 'mouseenter');
            });
    }

    /**
     * Handles the scroll event on the scrollable code content element and updates the header shadow accordingly.
     */
    private handleScroll(): void {
        this.scrollableCodeContent
            .elementScrolled()
            .pipe(takeUntilDestroyed(this.destroyRef))
            .subscribe(() => {
                this.setupHeaderShadow(this.scrollableCodeContent.measureScrollOffset('top') > 0);
            });
    }

    /**
     * Adds or removes the shadow class from the code block based on the specified condition.
     *
     * @param shouldShowShadow - A boolean indicating whether to add or remove the shadow class.
     */
    private setupHeaderShadow(shouldShowShadow: boolean): void {
        const className = 'kbq-code-block_header-with-shadow';

        if (shouldShowShadow) {
            this.renderer.addClass(this.elementRef.nativeElement, className);
        } else {
            this.renderer.removeClass(this.elementRef.nativeElement, className);
        }
    }

    /**
     * Adds or removes the actionbar display class from the code block based on the specified condition.
     *
     * The actionbar is always visible on mobile devices and when the tabs are visible.
     * Otherwise, the actionbar is only visible when the mouse is hovered over the code block.
     *
     * @param shouldShowActionbar - A boolean indicating whether the actionbar should be visible.
     */
    private setupActionbarDisplay(shouldShowActionbar?: boolean): void {
        const className = 'kbq-code-block_show-actionbar';

        // Should always show actionbar on Mobile devices
        if (this.platform.IOS || this.platform.ANDROID) {
            this.renderer.addClass(this.elementRef.nativeElement, className);
            return;
        }

        // Should always show actionbar when tabs are visible
        if (!this.hideTabs) {
            this.renderer.addClass(this.elementRef.nativeElement, className);
            return;
        }

        if (typeof shouldShowActionbar === 'undefined') {
            return;
        }

        if (shouldShowActionbar) {
            this.renderer.addClass(this.elementRef.nativeElement, className);
        } else {
            this.renderer.removeClass(this.elementRef.nativeElement, className);
        }
    }

    /** Whether the element has scroll. */
    private hasScroll({ scrollHeight, scrollWidth, clientHeight, clientWidth }: HTMLElement): boolean {
        return scrollHeight > clientHeight || scrollWidth > clientWidth;
    }

    /** Handles the enter keydown event on `viewAll` button. */
    protected onViewAllEnterKeydown(): void {
        this.toggleViewAll();
        if (this.canCodeContentBeFocused) {
            this.focusMonitor.focusVia(this.scrollableCodeContent.getElementRef().nativeElement, 'keyboard');
        }
    }

    private updateLocaleParams = (): void => {
        this._localeConfiguration = this.localeService?.getParams('codeBlock');

        this.changeDetectorRef.markForCheck();
    };

    /**
     * Copies the file code to the clipboard.
     *
     * If the copy was successful, the copy button tooltip content is updated
     * to show the "copied" message.
     */
    protected copyCode(): void {
        const file = this.files[this.activeFileIndex];
        if (this.clipboard.copy(file.content) && this.copyButtonTooltip) {
            this.copyButtonTooltip.content = this.localeConfiguration.copiedTooltip;
        }
    }

    /**
     * Opens the file link in a new window.
     */
    protected openLink(): void {
        const window = this.getWindow();
        const file = this.files[this.activeFileIndex];
        const safeURL = this.domSanitizer.sanitize(SecurityContext.URL, file.link!);
        if (safeURL) {
            window?.open(safeURL.toString(), '_blank');
        }
    }

    /**
     * Downloads the file as a blob.
     *
     * Creates a link with a blob as href and the file name as download attribute.
     * Then simulates a click event on the link to initiate the download.
     */
    protected downloadCode(): void {
        const file = this.files[this.activeFileIndex];
        const blob = new Blob([file.content], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const link = this.document.createElement('a');
        link.setAttribute('href', url);
        link.setAttribute('download', file.filename || this.fallbackFileName);
        link.click();
    }

    /** Use defaultView of injected document if available or fallback to global window reference */
    private getWindow(): Window | null {
        return this.document?.defaultView || window;
    }
}
