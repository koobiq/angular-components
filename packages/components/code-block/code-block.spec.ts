import { Component, ViewChild } from '@angular/core';
import { ComponentFixture, TestBed, fakeAsync, tick } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { HighlightModule } from 'ngx-highlightjs';
import { codeHTML, codeXML } from '../../components-dev/code-block/code-files-example';
import { KbqCodeBlockComponent } from './code-block.component';
import { KbqCodeBlockModule } from './code-block.module';

describe('CodeBlockComponent', () => {
    let component: KbqCodeBlockDefault;
    let fixture: ComponentFixture<KbqCodeBlockDefault>;

    beforeEach(() => {
        TestBed.configureTestingModule({
            imports: [
                NoopAnimationsModule,
                HighlightModule,
                KbqCodeBlockModule
            ],
            declarations: [KbqCodeBlockDefault]
        }).compileComponents();

        fixture = TestBed.createComponent(KbqCodeBlockDefault);
        component = fixture.componentInstance;

        fixture.detectChanges();
    });

    it('should create', () => {
        expect(component.codeBlock).toBeTruthy();
    });

    it('should switch value on expanding code block', () => {
        component.codeBlock.toggleViewAll();

        expect(component.codeBlock.viewAll).toBe(true);
    });

    it('should have less-contrast class', () => {
        const codeContainer = fixture.debugElement.query(By.css('.kbq-code-block'));
        const codeContainerEl: HTMLElement = codeContainer.nativeElement;

        expect(codeContainerEl.classList.contains('kbq-code-block_filled')).toBeTruthy();
    });

    it('should have 2 tabs', () => {
        const tabsGroup = fixture.debugElement.queryAll(By.css('.kbq-tab-label'));

        expect(tabsGroup.length).toBe(2);
    });

    it('should invoke handler methods after click on buttons in action bar', fakeAsync(() => {
        const codeButtons = fixture.debugElement.queryAll(By.css('.kbq-code-block-actionbar .kbq-button-icon'));
        const downloadCodeSpyFn = jest.spyOn(component.codeBlock, 'downloadCode');
        const copyCodeSpyFn = jest.spyOn(component.codeBlock, 'copyCode');
        const openExternalSystemSpyFn = jest.spyOn(component.codeBlock, 'openExternalSystem');

        expect(codeButtons.length).toBe(3);

        codeButtons[0].triggerEventHandler('click');
        codeButtons[1].triggerEventHandler('click');
        tick(100);
        codeButtons[2].triggerEventHandler('click');

        expect(downloadCodeSpyFn).toHaveBeenCalledTimes(1);
        expect(copyCodeSpyFn).toHaveBeenCalledTimes(1);
        expect(openExternalSystemSpyFn).toHaveBeenCalledTimes(1);
    }));
});

@Component({
    template: `
        <kbq-code-block [filled]="lessContrast" [codeFiles]="codeFiles" [maxHeight]="200" />
    `
})
class KbqCodeBlockDefault {
    @ViewChild(KbqCodeBlockComponent) codeBlock: KbqCodeBlockComponent;

    codeFiles = [
        {
            filename: 'index.html',
            content: codeHTML,
            language: 'html',
            link: 'https://stackblitz.com/edit/web-platform-f5jywg?file=index.html'
        },
        {
            filename: 'menu.xml',
            content: codeXML,
            language: 'xml',
            link: 'https://stackblitz.com/edit/web-platform-f5jywg?file=index.html'
        }
    ];

    lessContrast = true;
}
