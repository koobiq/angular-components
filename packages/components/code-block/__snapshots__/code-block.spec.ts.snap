// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KbqCodeBlock should provide custom locale configuration 1`] = `
{
  "copiedTooltip": "*unit_test* âœ“ Copied",
  "copyTooltip": "*unit_test* Copy",
  "downloadTooltip": "*unit_test* Download",
  "openExternalSystemTooltip": "*unit_test* Open in the external system",
  "softWrapOffTooltip": "*unit_test* Disable word wrap",
  "softWrapOnTooltip": "*unit_test* Enable word wrap",
  "viewAllText": "*unit_test* Show all",
  "viewLessText": "*unit_test* Show less",
}
`;
